### ISSUE TITLE
Flux Datasource: v.windowPeriod is replaced by 0s instead of an actual period by alerting engine

### ISSUE THREAD

**Issue by coussej** [grafana/grafana#26695]
**What happened**:
In a flux query, `v.windowPeriod` gets replaced with the window interval of the current selected period in Grafana, just like the old `$__interval` variable. In the exact same query, executed by the alerting on the backend, it gets replaced by `0s`, resulting in an error.

<img width="332" alt="image" src="https://user-images.githubusercontent.com/9944663/88919141-e646fe00-d26a-11ea-9019-3b21be448a38.png">

**What you expected to happen**:
`v.windowPeriod` to be replaced with the window interval of the configured alert period.

**How to reproduce it (as minimally and precisely as possible)**:
Configure a query using this variable an set an alert on it.

**Anything else we need to know?**:
The actual queries are logged by grafana-server. 

This is when executing from the query editor:
<img width="434" alt="image" src="https://user-images.githubusercontent.com/9944663/88919300-260de580-d26b-11ea-89b6-fca70d8cee02.png">

This is when executing from the 'test alert rule':
<img width="417" alt="image" src="https://user-images.githubusercontent.com/9944663/88919277-1b535080-d26b-11ea-9e7d-5d750df3bb8e.png">

**Environment**:
- Grafana version: 7.1.1 
- Data source type & version: Flux (integrated in the InfluxDB source).
- OS Grafana is installed on: Ubuntu
- User OS & Browser: Not relevant
- Grafana plugins: /
- Others: Influxdb 1.8.1

**Comment by aknuds1** [comment]
Is this version 7.1.1?

**Comment by coussej** [comment]
Yes, 7.1.1

**Comment by MatinF** [comment]
I'm struggling to get Grafana 7.2 + Flux to work with alerts as well. I've added my case below in this thread (my other thread was closed because it was considered a duplicate of this one):

---------

I'm using Grafana cloud (7.1.5) with InfluxDB cloud 2.0 (Flux). 

I'm trying to set up a basic alert in Grafana, which should trigger based on an InfluxDB based time series chart.

**My default Flux query**
My normal Flux queries look as below, which produce the expected result in my chart:
```
from(bucket: v.defaultBucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "3F78A21D")
  |> filter(fn: (r) => r["_field"] == "S1_PID_0D_VehicleSpeed")
  |> aggregateWindow(every: v.windowPeriod, fn: mean)
  |> yield(name: "mean")
```
![image](https://user-images.githubusercontent.com/26184427/93451234-32abd300-f8d7-11ea-92fd-a9191940029f.png)


**The problem**
However, when I try to test an alert based on this I get the below error:
```
error:"tsdb.HandleRequest() response error &{invalid: error calling function "aggregateWindow" @5:6-5:42: error calling function "window" @universe.flux|125:12-125:57: parameter "every" must be nonzero A <nil> [] [] 0xc001512c00}"
```
**Attempted solution**
Based on the error, I've tried removing the Window aggregation, resulting in the below query:
```
from(bucket: v.defaultBucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "3F78A21D")
  |> filter(fn: (r) => r["_field"] == "S1_PID_0D_VehicleSpeed")
  |> yield(name: "mean")
```
![image](https://user-images.githubusercontent.com/26184427/93451349-5bcc6380-f8d7-11ea-8d01-14e67d9ee6b8.png)

When I test the alert rule with this query, I do not get an error. However, if I use this adjusted query, my chart only displays part of my data series (approximately 1 hour of data is not shown), as evident from the above picture. 

Am I missing something basic in setting this up, or is there a bug in the Flux alert functionality? I was not able to find specific examples of Flux queries in the context of setting up alerts. Basically I would like to be able to add alerts to charts that I also use for visualization in my dashboard.

**Comment by coussej** [comment]
@MatinF the flux datasource takes into account the "Max data points" of the query options. Try increasing that for your graph to show up completely.

![image](https://user-images.githubusercontent.com/9944663/94905099-9cfa7100-049c-11eb-862d-6765faa0d196.png)

**Comment by MatinF** [comment]
@coussej Thanks for the inputs. The challenge with boosting the max data points is that my charts are at 1 second resolution. As such, if I were to be able to include all data points this way, it would result in a huge number of datapoints if a user e.g. views the chart for a period of 1 week.

The original query I shared solves this by aggregation, which works perfectly for visualizing my data. 

I guess what I'm struggling with is this: Is there a good way to be able to use Flux in Grafana for both creating a high performant panel (ala my original query) - while at the same time enabling the use of alerts via the same panel? 

If no, is the "intended method" for Flux Grafana alerts that the user sets up separate panels for the alerts, e.g. hiding these from view? If so, what would the query ideally look like for such a panel - and is it necessary to set the Max data points to a high value in order for the alerts to work, or is that purely relevant for the visual display?

**Comment by coussej** [comment]
@MatinF my opinion is that as long as this issue with the v.windowPeriod is nog solved for alerting, the high-performant panel with alerting combo is not really possible, as you need to fix the interval for the alerting query to work.

**Comment by MatinF** [comment]
Ok if so, my understanding is that there's no "proper way" to set up alerts in Grafana with InfluxDB Flux currently. Would be really great to get this resolved as it is a critical functionality for a lot of the use cases our end users encounter. If there's a work around for it short-term, I'd also be interested in that.

**Comment by ryantxu** [comment]
See also https://github.com/grafana/grafana/issues/27469

7.2 has a fix that sets the maxDataPoints and interval  https://github.com/grafana/grafana/pull/27463 -- but you may have other issues

**Comment by MatinF** [comment]
Thanks, I'm on 7.2 (via the Cloud version of Grafana), but I still struggle to see how I should set up a panel alert via Flux in the "correct" way. Is there some example of a Flux query that lends itself well to setting up alerts as well?

To recap, the query I use currently looks as follows:
```
from(bucket: v.defaultBucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "3F78A21D")
  |> filter(fn: (r) => r["_field"] == "S1_PID_0D_VehicleSpeed")
  |> aggregateWindow(every: v.windowPeriod, fn: mean)
  |> yield(name: "mean")
```

With alert settings as below:
![image](https://user-images.githubusercontent.com/26184427/95022109-6fb4eb00-0675-11eb-9f51-9d1b20558afa.png)

This does not seem to work, however, and I am not able to find any clear guides to see how I should be setting this up with Flux in v7.2. Any inputs would be really appreciated.

**Comment by MatinF** [comment]
Any updates on whether this might be resolved via an update to Grafana in the near future? We've got a lot of Grafana users that we support who would need this functionality.

**Comment by darwin10** [comment]
example workaround 
windowPeriod = int(v: v.windowPeriod)
myWindow = if windowPeriod < int(v: 100000000) then 1s else v.windowPeriod
from(bucket: "RottenKids")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "inside/closet/temphum1" )
  |> filter(fn: (r) => r["_field"] == "temperature" )
  |> aggregateWindow(every: myWindow, fn: mean)
  |> yield(name: "mean")

**Comment by MatinF** [comment]
Hi Darwin, thanks for the input - can you clarify in what way this resolves the issue, just to understand it?

Martin

**Comment by darwin10** [comment]
the window period set during the alert is too small or zero, this checks and overrides that minimum value with 1s, allowing the alert to work.

**Comment by MatinF** [comment]
Ok thanks I'll try it out - would the other settings I showed for the alert itself be compatible with this solution, or will my alert rules need to be modified for it to work as expected?

**Comment by darwin10** [comment]
Everything else should work.

**Comment by MatinF** [comment]
OK thanks a lot, I'll try it out

**Comment by gelldur** [comment]
Bug still present in `v7.4.3 (010f20c1c8)`, workaround works.

**Comment by gabor** [comment]
we are planning some changes in alerting in the near future that could change/fix this, so we will look at this again after those changes happen.

**Comment by pheetr** [comment]
I haven't seen one thing mentioned, namely how this impacts InfluxDB's performance/stability.

Granted, my implementation was fairly lightweight, with one vCPU and 2GB of RAM, but this was more than enough for my TIG stack monitoring about a dozen devices and services. A day after I've done some tuning of the panel queries I noticed that the entire VM had become unresponsive. It took me about a day of troubleshooting to finally narrow it down to v.windowPeriod I had been so happy to have discovered a day or so before. This was now present in pretty much all my graph panel queries, including the two that had alerts configured.

Some stats about how this impacts InfluxDB
Without v.windowPeriod in an alerting query:
![image](https://user-images.githubusercontent.com/44616229/114266216-08457280-99ed-11eb-9561-8031f6116912.png)
After an alert's "Test rule" was triggered a single time:
![image](https://user-images.githubusercontent.com/44616229/114266353-cbc64680-99ed-11eb-88eb-37a26379c746.png)

What is not visible on the screenshots is that these queries run for 30sec before timing out and returning an error. During these 30sec the memory allocation of InfluxDB increases rapidly and its CPU utilization is at 100%.
The unresponsiveness could be avoided by providing a lot more resources (primarily memory) to the system. As part of the troubleshooting I've observed that with two alerts configured with evaluation every 1m the memory allocation of InfluxDB finally settled at slightly over 10GB with a more or less permanent 90+% CPU utilization.
I've tested with InfluxDB 2.0.4 and Grafana 7.5.3.

I'm looking forward to this issue being resolved, hopefully near-term. 

In the meantime though, I would suggest adding a warning about this issue to the [Supported macros section of the InfluxDB datasource page](https://grafana.com/docs/grafana/latest/datasources/influxdb/#supported-macros). 
This is where I've learned about the v.windowPeriod macro as well and I believe it would be a great place to help others avoid running into the same issue and spending time trying to figure out what's going wrong unnecessarily.

Thanks.
